crates/sui-indexer/src/indexer_reader.rs

```diff
@@ -633,58 +653,128 @@ impl IndexerReader {
+                )
             }
             Some(TransactionFilter::FromAndToAddress { from, to }) => {
                 let from_address = Hex::encode(from.to_vec());
                 let to_address = Hex::encode(to.to_vec());
-                format!(
-                    "(senders @> ARRAY['\\x{}'::bytea] AND recipients @> ARRAY['\\x{}'::bytea])",
-                    from_address, to_address
-                )
+                // Need to de-amibguize the tx_sequence_number column
+                let cursor_clause = if let Some(cursor_tx_seq) = cursor_tx_seq {
+                    if is_descending {
+                        format!(
+                            "AND tx_senders.{TX_SEQUENCE_NUMBER_STR} < {}",
+                            cursor_tx_seq
+                        )
+                    } else {
+                        format!(
+                            "AND tx_senders.{TX_SEQUENCE_NUMBER_STR} > {}",
+                            cursor_tx_seq
+                        )
+                    }
+                } else {
+                    "".to_string()
+                };
+                let inner_query = format!(
+                    "(SELECT tx_senders.{TX_SEQUENCE_NUMBER_STR} \
+                    FROM tx_senders \
+                    JOIN tx_recipients \
+                    ON tx_senders.{TX_SEQUENCE_NUMBER_STR} = tx_recipients.{TX_SEQUENCE_NUMBER_STR} \
+                    WHERE tx_senders.sender = '\\x{}'::BYTEA \
+                    AND tx_recipients.recipient = '\\x{}'::BYTEA \
+                    {} \
+                    ORDER BY {TX_SEQUENCE_NUMBER_STR} {} \
+                    LIMIT {}) AS inner_query
+                    ",
+                    from_address,
+                    to_address,
+                    cursor_clause,
+                    order_str,
+                    limit,
+                );
+                (inner_query, "1 = 1".into())
             }
             Some(TransactionFilter::FromOrToAddress { addr }) => {
                 let address = Hex::encode(addr.to_vec());
-                format!(
-                    "(senders @> ARRAY['\\x{}'::bytea] OR recipients @> ARRAY['\\x{}'::bytea])",
-                    address, address
-                )
+                let inner_query = format!(
+                    "( \
+                        ( \
+                            SELECT {TX_SEQUENCE_NUMBER_STR} FROM tx_senders \
+                            WHERE sender = '\\x{}'::BYTEA {} \
+                            ORDER BY {TX_SEQUENCE_NUMBER_STR} {} \
+                            LIMIT {} \
+                        ) \
+                        UNION \
+                        ( \
+                            SELECT {TX_SEQUENCE_NUMBER_STR} FROM tx_recipients \
+                            WHERE recipient = '\\x{}'::BYTEA {} \
+                            ORDER BY {TX_SEQUENCE_NUMBER_STR} {} \
+                            LIMIT {} \
+                        ) \
```

Comment:
the `UNION` here might include duplicate txs, and directly FULL JOIN will do seq scan again when I try it locally, but we can do query like
```
EXPLAIN WITH l as (SELECT ts.tx_sequence_number FROM tx_senders ts WHERE ts.sender = 'A'::BYTEA LIMIT 100),
r as (SELECT tr.tx_sequence_number FROM tx_recipients tr WHERE tr.recipient = 'A'::BYTEA LIMIT 100)
SELECT l.tx_sequence_number FROM l FULL JOIN r ON l.tx_sequence_number = r.tx_sequence_number LIMIT 100;
                                                    QUERY PLAN
------------------------------------------------------------------------------------------------------------------
 Limit  (cost=6.49..14.53 rows=1 width=8)
   ->  Hash Full Join  (cost=6.49..14.53 rows=1 width=8)
         Hash Cond: (tr.tx_sequence_number = ts.tx_sequence_number)
         ->  Limit  (cost=0.28..8.29 rows=1 width=8)
               ->  Index Only Scan using tx_recipients_pkey on tx_recipients tr  (cost=0.28..8.29 rows=1 width=8)
                     Index Cond: (recipient = '\x41'::bytea)
         ->  Hash  (cost=6.20..6.20 rows=1 width=8)
               ->  Limit  (cost=0.42..6.19 rows=1 width=8)
                     ->  Index Only Scan using tx_senders_pkey on tx_senders ts  (cost=0.42..6.19 rows=1 width=8)
                           Index Cond: (sender = '\x41'::bytea)
(10 rows)
```

Reply:
`UNION` dedups rows: https://www.postgresql.org/docs/current/queries-union.html